src/apps/js_generic/js_generic.cpp

```diff
@@ -781,6 +782,114 @@ namespace ccfapp
       JS_FreeValue(ctx, global_obj);
     }
 
+    static JSValue create_json_obj(const nlohmann::json& j, JSContext* ctx)
+    {
+      const auto buf = j.dump();
+      return JS_ParseJSON(ctx, buf.data(), buf.size(), "<json>");
+    }
+
+    static JSValue create_caller_obj(EndpointContext& args, JSContext* ctx)
+    {
+      if (args.caller == nullptr)
+      {
+        return JS_NULL;
+      }
+
+      auto caller = JS_NewObject(ctx);
+
+      if (auto jwt_ident = args.try_get_caller<ccf::JwtAuthnIdentity>())
+      {
+        JS_SetPropertyStr(
+          ctx,
+          caller,
+          "policy",
+          JS_NewString(ctx, get_policy_name_from_ident(jwt_ident)));
+
+        auto jwt = JS_NewObject(ctx);
+        JS_SetPropertyStr(
+          ctx,
+          jwt,
+          "key_issuer",
+          JS_NewStringLen(
+            ctx, jwt_ident->key_issuer.data(), jwt_ident->key_issuer.size()));
+        JS_SetPropertyStr(
+          ctx, jwt, "header", create_json_obj(jwt_ident->header, ctx));
+        JS_SetPropertyStr(
+          ctx, jwt, "payload", create_json_obj(jwt_ident->payload, ctx));
+        JS_SetPropertyStr(ctx, caller, "jwt", jwt);
+
+        return caller;
+      }
+      else if (
+        auto empty_ident = args.try_get_caller<ccf::EmptyAuthnIdentity>())
+      {
+        JS_SetPropertyStr(
+          ctx,
+          caller,
+          "policy",
+          JS_NewString(ctx, get_policy_name_from_ident(empty_ident)));
+        return caller;
+      }
+
+      char const* policy_name = nullptr;
+      size_t id;
+      nlohmann::json data;
+      std::string cert_s;
+
+      if (
```

Comment:
It would have seemed more natural to follow the order given by the auth list here, in case some of the values end up overwriting each other. Either or separate the output of each policy so they can be fetched separately.

If I specify [sig, cert], and they yield different identities, I will get the cert identity, not because I specified it last but because of this hidden hardcoded order. That doesn't seem right? Am I missing something?

Reply:
Once the first policy passes, we won't execute any others, so (at-most) one identity is set. Only one of these branches can be executed. The priority is determined by the order the policies are listed, either in `make_endpoint()` or `app.json`.

If you specify `[sig, cert]`, and the sig is valid, you'll get the sig identity. If there's no sig but a valid cert, you'll get the cert identity.